frame/dapp-staking-v3/src/lib.rs

```diff
@@ -0,0 +1,730 @@
+// This file is part of Astar.
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+//! # dApp Staking v3 Pallet
+//!
+//! - [`Config`]
+//!
+//! ## Overview
+//!
+//! Pallet that implements dapps staking protocol.
+//!
+//! <>
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Function
+//!
+//! <>
+//!
+//! ### Other
+//!
+//! <>
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+    pallet_prelude::*,
+    traits::{Currency, LockIdentifier, LockableCurrency, StorageVersion, WithdrawReasons},
+    weights::Weight,
+    BoundedVec,
+};
+use frame_system::pallet_prelude::*;
+use parity_scale_codec::{Decode, Encode};
+use sp_runtime::traits::{AtLeast32BitUnsigned, BadOrigin, Saturating, Zero};
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod test;
+
+const STAKING_ID: LockIdentifier = *b"dapstake";
+
+// TODO: move all structs & types into a separate `types.rs` file, after receiving first review batch
+
+/// The balance type used by the currency system.
+pub type BalanceOf<T> =
+    <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+/// Convenience type for `AccountLedger` usage.
+pub type AccountLedgerFor<T> = AccountLedger<
+    BalanceOf<T>,
+    BlockNumberFor<T>,
+    <T as Config>::MaxLockedChunks,
+    <T as Config>::MaxUnlockingChunks,
+>;
+
+/// Era number type
+pub type EraNumber = u32;
+/// Period number type
+pub type PeriodNumber = u32;
+/// Dapp Id type
+pub type DAppId = u16;
+
+/// Distinct period types in dApp staking protocol.
+#[derive(Encode, Decode, MaxEncodedLen, Clone, Copy, Debug, PartialEq, Eq, TypeInfo)]
+pub enum PeriodType {
+    /// Period during which the focus is on voting.
+    /// Inner value is the era in which the voting period ends.
+    VotingPeriod(#[codec(compact)] EraNumber),
```

Comment:
```suggestion
    Voting {
        #[codec(compact)]
        ends_at: EraNumber,
    }),
```

It is more about personal preference, but I think this is more readable and self-explanatory. `PeriodType` naming includes `Period` space so we don't need to repeat.

If it makes sense, we can apply this to other enum fields.

Reply:
I agree 😁 

Haven't gotten to the part of implementation where I use this, but I'd probably get the same idea as you after writing down `PeriodType::VotingPeriod` xD